### ISSUE TITLE
Lock status when fail to spawn pppd in macOS

### ISSUE THREAD

**Issue by Mabin-J** [adrienverge/openfortivpn#137]
This issue for bind PR.

----

When I run openfortivpn without elevate privileges in Linux,
it is finished normally but not in macOS.

It showed `INFO:   Cancelling threads...` and locked.
It didn't response when I push `Ctrl + C`

This is full message.
```
$ ./openfortivpn -c connect_info.config -v
DEBUG:  Loaded config file "connect_info.config".
VPN account password: 
DEBUG:  Config host = "123.123.123.123"
DEBUG:  Config realm = ""
DEBUG:  Config port = "443"
DEBUG:  Config username = "username"
DEBUG:  Config password = "********"
WARN:   This process was not spawned with root privileges, this will probably not work.
DEBUG:  Gateway certificate validation failed.
DEBUG:  Gateway certificate digest found in white list.
INFO:   Connected to gateway.
INFO:   Authenticated.
DEBUG:  Cookie: SVPNCOOKIE=[COOKIE]
INFO:   Remote gateway has allocated a VPN.
DEBUG:  Gateway certificate validation failed.
DEBUG:  Gateway certificate digest found in white list.
WARN:   No gateway address, using interface for routing
DEBUG:  pppd_read_thread
DEBUG:  ssl_read_thread
DEBUG:  ssl_write_thread
DEBUG:  if_config thread
DEBUG:  gateway ---> pppd (12 bytes)
DEBUG:  pppd_write thread
DEBUG:  pppd ---> gateway (12 bytes)
WARN:   read returned 0
WARN:   read returned 0
WARN:   read returned 0

(repeat about 940 times...)

WARN:   read returned 0
WARN:   read returned 0
WARN:   read returned 0
DEBUG:  gateway ---> pppd (12 bytes)
WARN:   read returned 0
WARN:   read returned 0
WARN:   read returned 0
WARN:   read returned 0
WARN:   read returned 0
WARN:   read returned 0
WARN:   read returned 0
WARN:   read returned 0
WARN:   read returned 0
WARN:   read returned 0
WARN:   read returned 0
WARN:   read returned 0
ERROR:  write: Input/output error
WARN:   read returned 0
INFO:   Cancelling threads...
WARN:   read returned 0

(Not response)

(push Ctrl + C)
^C^C 

(Not response)

(run `kill pid` in another terminal)

[1]    2201 terminated  ./openfortivpn -c connect_info.config -v
```

I checked something, and I found that `pppd` zombie process...
```
$ ps aux |grep ppp
username          2286   0.0  0.0        0      0   ??  Z     2:22PM   0:00.00 (pppd)
username          2288   0.0  0.0  2432804    780 s005  S+    2:22PM   0:00.00 grep --color=auto --exclude-dir=.bzr --exclude-dir=CVS --exclude-dir=.git --exclude-dir=.hg --exclude-dir=.svn ppp
```
The 2286 process was remained until openfortivpn was killed.

**Comment by Mabin-J** [comment]
This issue is solved with #132 but it is temporary 
because I didn't find core cause at that time.

And my guess is wrong at that time, too.
@mrbaseman posts #133 and I wrote comment about #132.

I found why openfortivpn is locked. so I push PR #135

**Comment by mrbaseman** [comment]
@Mabin-J Thanks for explaining again. I have read a bit more in the references (thanks to @DimitriPapadopoulos for the links) and studied #135. I have posted a few comments specific to this change in that thread. 

@adrienverge since the patch immediately reverts the most recent commit on master, should we squash the two pull requests into one commit after the merge? The history of the repository would be cleaner then. An argument against would be that some (unexperienced) git users might have already cloned the current master and might have problems when pulling the squashed commit.

Another point: #105 is perhaps also related to the current issue, but I'm not sure how to deal with that one.

**Comment by mrbaseman** [comment]
@DimitriPapadopoulos if my understanding is correct now, #134 is an independent story (although discussion continued there), right? 
So, if we decide to squash it would make sense to land #134 afterwards.

**Comment by DimitriPapadopoulos** [comment]
Yes, #134 is a different story, orthogonal to this one.

I can rebase #134 afterwards if needed, no worries.

**Comment by DimitriPapadopoulos** [comment]
@mrbaseman Yes, #105 certainly does seem related. I don't have a Mac and I don't know much about Mac OS programming, so it's hard for me to tell whether this is the exact same issue or not. Would need to read some documentation first...

**Comment by adrienverge** [comment]
> since the patch immediately reverts the most recent commit on master, should we squash the two pull requests into one commit after the merge? The history of the repository would be cleaner then.

I understand the idea, but I'd prefer we keep the convention to only push *fast-forward* changes to the special branch `master` (which means never `push --force`). So it won't be clean... but there is a good reason :)

**Comment by mrbaseman** [comment]
ok, so let's stay with the policy of never `push --force` to master. In this case it doesn't matter which of the currently open pull requests is landed and which ones have to be rebased then. Let's see where we first agree for landing.

**Comment by mrbaseman** [comment]
I have just merged #135 so we can also close this issue for the general discussion (which anyway continued in #105)
